opal/corelib/kernel.rb

```diff
@@ -779,16 +779,33 @@ def instance_variable_get(name)
   def instance_variable_set(name, value)
     name = Opal.instance_variable_name!(name)
 
-    `self[name.substr(1)] = value`
+    `self[Opal.ivar(name.substr(1))] = value`
   end
 
+  %x{
+    function restoreIvar(name) {
+      if (name === "constructor$" ||
+        name === "__proto__$" ||
+        name === "__parent__$" ||
+        name === "__noSuchMethod__$" ||
+        name === "__count__$" ||
+        name === "hasOwnProperty$" ||
+        name === "valueOf$")
+      {
+        return name.slice(0, name.length - 1);
+      }
+
+      return name;
+    }
+  }
+
```

Comment:
@iliabylich I see what you mean, I think we can ignore that check as we already look for all own properties that match a valid ivar name. Also MRI doesn't allow for `$` to be in an ivar name:

```
irb(main):001:0> o = Object.new
=> #<Object:0x007fad74c7c530>
irb(main):002:0> o.instance_variable_set '@asdf$', 123
NameError: `@asdf$' is not allowed as an instance variable name
	from (irb):2:in `instance_variable_set'
	from (irb):2
	from /Users/elia/.rvm/rubies/ruby-2.2.1/bin/irb:11:in `<main>'
irb(main):003:0> 
```

@vais good point, this should be better: `return name.substr(-1) === '$' ? name.slice(0, name.length-1) : name`